### PR TITLE
Improve input mappings with conflict handling and unbound warnings

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

image

image

image

image

image

## Description

What does this PR do? (e.g., "Fixes player jump physics in level 2" or "Adds
new enemy AI script")

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [x] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [x] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [x] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [x] Any new unit tests added? (Link to test scene if yes)
- [x] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [x] Code follows Godot style guide (e.g., snake_case for variables)
- [x] No console errors in editor/output
- [x] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Enhancements:
- Add a post-load check that backfills missing default keyboard and gamepad mappings into the InputMap and persists them to settings when needed.

## Summary by Sourcery

Improve input mapping robustness, conflict handling, and device-aware UI/UX across settings, key mapping, main menu, and gameplay while expanding automated tests and documentation.

New Features:
- Add conflict detection and confirmation dialog when assigning inputs already used by other actions, with device-specific unbinding logic.
- Introduce per-device tracking and persistence of the last selected input device to drive labels, prompts, and unbound checks.
- Display in-game warnings for unbound critical controls or presses of completely unbound inputs, guiding players to the key mapping menu.
- Provide helper APIs for querying pause binding labels, event bindings, conflicts, and event device types from settings.
- Add a utility to load the Key Mapping menu directly from other menus while preserving the background video and menu stack.
- Extend main scene messaging system to show temporary HUD messages with typed flags for different warning types.

Bug Fixes:
- Ensure default input mappings are backfilled and saved robustly even with corrupted or legacy config formats, including proper handling of explicit unbound states.
- Fix migration logic to use a generic save-needed flag instead of an old migration-only flag and make legacy unbound critical actions migrated once and persisted.
- Prevent gamepad defaults or resets from creating duplicate or cross-action bindings, fully clearing device-specific events before restoring defaults.
- Avoid crashes and misbehavior when loading malformed input arrays by ignoring non-string entries and treating invalid data safely.
- Make loading screen progress rely on real threaded loading progress (with web-specific timing) to avoid the progress bar getting stuck or desynced.
- Ensure focus restoration when exiting the key mapping menu returns to the appropriate button in options or main menu, improving keyboard/DPAD navigation.
- Fix unbound critical-controls warning so it is device-aware and shown only once per session until bindings are actually fixed.

Enhancements:
- Refactor input default backfilling into a shared helper for reuse in settings, migrations, and tests.
- Tighten reset-to-default behavior to log and fully reset per device while keeping other device bindings intact.
- Refine remap button prompts to be device-specific and adjust remap flow to support conflict-aware behavior and better logging.
- Improve key mapping menu focus logic so keyboard/gamepad toggles, remap buttons, and back/reset controls are reachable via keyboard/DPAD from first open.
- Expand tests for settings, input remapping, key mapping UI, and combined config loading to cover new conflict, unbound, migration, and device-state behaviors.
- Clarify and slightly reorganize global utility scripts (globals.gd, loading_screen.gd, main_scene.gd) with better comments and small structural improvements.

Documentation:
- Add a deployment guide detailing free web browser game platforms and recommended automation strategies for Godot Web exports.
- Update README to reference the new web deployment platforms guide.

Tests:
- Add new edge-case tests for settings to cover conflict-unbound persistence, last-device validation, legacy migration, and corrupt config recovery.
- Extend key mapping menu tests to cover improved device-aware labels, reset behavior, persistence, and default backfill for keyboard and gamepad.
- Update input remap button tests to validate device-specific prompts, direct remap behavior, and no-input cancellation after UI changes.
- Adjust existing system tests to align with the new save flag semantics and per-device default behavior for inputs.

Chores:
- Tidy comments and headers across multiple scripts to better document responsibilities and link to external test plans.

## Summary by Sourcery

Improve input handling, device awareness, and user feedback for key mappings and gameplay while hardening settings migration and loading behavior.

New Features:
- Add conflict-aware input remapping with a confirmation dialog that unbinds only conflicting device-specific bindings.
- Introduce device-aware pause binding labels and last-used input device persistence to drive UI labels and warnings.
- Display in-game HUD messages for critical unbound controls and presses of entirely unbound inputs, guiding players toward the key mapping menu.
- Provide global helper to open the Key Mapping menu directly from other menus while keeping the background video playing.

Bug Fixes:
- Ensure default input mappings are backfilled and saved even when configs are missing, corrupted, or in legacy formats, including explicit unbound states.
- Prevent reset-to-default and gamepad defaults from creating duplicate or cross-action bindings by fully clearing per-device events before restoring defaults.
- Fix unbound critical-controls warning so it is device-aware, shown only once per session, and cleared when bindings are fixed.
- Avoid crashes when loading malformed input arrays by safely ignoring non-string entries and using robust type handling.
- Make the loading screen progress rely on actual threaded loading progress with web-specific timing to avoid stuck or misleading progress bars.
- Restore focus correctly when exiting the key mapping menu back to options or main menu to improve keyboard/controller navigation.

Enhancements:
- Refactor input default backfill and legacy migration into reusable helpers that respect explicit unbound states and per-device bindings.
- Add helpers for detecting conflicting actions, checking if events are bound, and determining event device type to support richer input logic.
- Improve key mapping menu focus behavior, device toggle handling, and remap prompts so keyboard and gamepad flows are clearer and more accessible.
- Centralize temporary HUD messaging with typed flags in the main scene to avoid repeated unbound warnings and manage message lifetimes cleanly.
- Switch global remap prompt text to device-specific variants to avoid layout shifts and clarify what input is expected.
- Clarify and slightly reorganize global utility scripts, loading screen logic, and options menu focus handling for better structure and consistency.

Documentation:
- Add a guide to free web browser deployment platforms for Godot Web exports and link it from the README.

Tests:
- Expand settings, input remapping, and key mapping menu tests to cover conflict handling, explicit unbound persistence, device-aware labels, and corrupt config recovery.
- Extend combined-config and error-edge-case tests to align with the new save flag semantics, legacy migration, and per-device default behavior.
- Add regression tests for pause binding labels, device validation, and HUD messages to ensure new input-label helpers behave as expected.

Chores:
- Tighten comments and headers across scripts and tests to better document responsibilities and reference external test plans.

## Summary by CodeRabbit

* **New Features**
  * Conflict-resolution dialog for control remapping, device-specific remap prompts, remember last input device, HUD messages for unbound inputs, Start Game confirmation when critical controls are unbound, and key-mapping submenu loader.

* **Bug Fixes**
  * Safer migration/backfill of input defaults, robust handling of corrupted legacy configs, fixes for duplicate bindings on reset, and reliable threaded loading progress.

* **Improvements**
  * Redesigned key-mapping navigation and focus, updated visuals and larger buttons, device-aware remap flow, improved back-navigation behavior.

* **Documentation**
  * Added guide for free web deployment platforms (CI/CD and export tips).

* **Tests**
  * Expanded and updated tests for remapping, conflict flows, device-aware prompts, and async UI sequencing.

### CODE REVIEW

**Comment by ikostan** [comment]
I found a new bug here: ame key is assigned to multiple actions with no warning or unbinding.

This is a classic “duplicate binding” issue in custom remapping systems, and the solution you described (warning → confirm → unbind the other) is exactly what modern games do (e.g. Doom, Cyberpunk, etc.). We’ll implement it cleanly, without touching the scene file at all (we create the dialog in code).

**Comment by ikostan** [comment]
On some devices we have an issue with the loading screen (for now detected in WEB version only, did not tested on PC clients). The problem is that it getting stuck on exactly 50% for a few seconds and than continues normally.

The 50% stuck issue is a classic Godot 4 quirk with ResourceLoader threaded loading—especially noticeable on Web exports.

**Comment by ikostan** [comment]
@sourcery-ai review

File: scripts/key_mapping.gd [hunk 1]
```diff
@@ -64,6 +84,84 @@ func _ready() -> void:
...
+
+	# Cleanup
+	_clear_conflict_state()
+
+	# NEW: Clear critical warning flag when player fixes unbound (once-per-session)
+	if Globals.get_node_or_null("/root/MainScene"):  # safe null check
+		var main: Node = Globals.get_node_or_null("/root/MainScene")
+		if main:
+			main._showing_unbound_warning = false
+
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/key_mapping.gd]: This one should be resolved, please review it again.
